ability to use lambda target in elb_target_group

markuman opened the pull request:
##### SUMMARY

Since late 2018 it is possible to use a lambda function as a target in a target group.  
This pull request add the ability to use a lambda function as a target.

##### ISSUE TYPE

- Feature Pull Request


##### COMPONENT NAME

elb_target_group

##### ADDITIONAL INFORMATION

This pull request fixes #52253 

```yml
---
- hosts: localhost
  gather_facts: False


  tasks:

    - name: first, create empty target group
      elb_target_group:
        name: my-lambda-targetgroup
        target_type: lambda
        state: present
        modify_targets: False
      register: out

    - name: second, allow invoke of the lambda
      lambda_policy:
        state: present
        function_name: mb_dev3
        statement_id: someID
        action: lambda:InvokeFunction
        principal: elasticloadbalancing.amazonaws.com
        source_arn: "{{ out.target_group_arn }}"

    - name: third, add target
      elb_target_group:
        name: my-lambda-targetgroup
        target_type: lambda
        state: present
        targets:
          - Id: arn:aws:lambda:eu-central-1:507463463146:function:mb_dev3
```


```
ansible-playbook test.yml 

PLAY [localhost] *******************************************************************************************************************************************************************************************************************

TASK [first, create empty target group] ********************************************************************************************************************************************************************************************
changed: [localhost]

TASK [second, allow invoke of the lambda] ******************************************************************************************************************************************************************************************
changed: [localhost]

TASK [third, add target] ***********************************************************************************************************************************************************************************************************
changed: [localhost]

PLAY RECAP *************************************************************************************************************************************************************************************************************************
localhost                  : ok=3    changed=3    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```

Commit:
make is basically works

Commit:
keep Id structure of targets. handle 'changed' status and implement deregister of a target

Commit:
add documentation about target lambda

Commit:
fix documentation.

Commit:
update example and resolve comment on code

Commit:
fix tailing whitespaces and example syntax

resmo reviewed (approved):
LGTM

willthames commented:
The change looks reasonable, but new features should have accompanying tests.

It looks like test/integration/targets/elb_target already tests elb_target_group, so would be good to add tests for this change to that test suite

willthames commented:
Alternatively you could add the test to the lambda test suite if it's easier, I don't really mind

jillr commented on lib/ansible/modules/cloud/amazon/elb_target_group.py:
```diff
@@ -110,14 +110,15 @@
-        Note that you can't specify targets for a target group using both instance IDs and IP addresses.
+        C(instance) (targets are specified by instance ID), C(ip) (targets are specified by IP address) or C(lambda) (target is specified by ARN).
+        Note that you can't specify targets for a target group using more than one type. Target type lambda only accept one target. When more than
+        one target is specified, only the first one is used. All follow targets are ignored.
```
```suggestion
        one target is specified, only the first one is used. All additional targets are ignored.
```

jillr commented on lib/ansible/modules/cloud/amazon/elb_target_group.py:
```diff
@@ -212,6 +213,37 @@
+# Using lambda as targets require that the target group
+# itself is allow to invoke the lambda function.
+# therefore you need first to create an empty target group
+# to receice its arn, second, allow the target group
```
```suggestion
# to receive its arn, second, allow the target group
```

jillr reviewed (commented):
Couple small docs nits, and would also like to see some tests.

markuman committed:
Update lib/ansible/modules/cloud/amazon/elb_target_group.py

Co-Authored-By: Jill R <4121322+jillr@users.noreply.github.com>

markuman committed:
Update lib/ansible/modules/cloud/amazon/elb_target_group.py

Co-Authored-By: Jill R <4121322+jillr@users.noreply.github.com>

markuman commented:
@willthames  jillr 
is it sufficient when the test does

1. create lambda function
2. create target group
3. allow invoke lambda function from elb
4. add lambda function as target to the target group
5. delete everything - even in case on errors


or do we need a test which also creates an ALB and make a test request?

jillr commented:
Those 5 items are good for me, with idempotency test for the actions.  For example;
4.a) add lambda function as target to the target group (idempotent)
assert: that: not result.changed
Thanks @markuman

Commit:
try to add tests

jillr commented on test/integration/targets/elb_target/playbooks/roles/elb_lambda_target/tasks/main.yml:
```diff
@@ -0,0 +1,101 @@
+---
+- name: set up lambda as elb_target
+
```
You need to setup the connection parameters and then use them in each of your tasks, like in:
https://github.com/ansible/ansible/blob/76392649308a51622808b79172c689a3ee41ed24/test/integration/targets/elb_target/playbooks/roles/elb_target/tasks/main.yml#L10-L26

markuman replied:
I fix the issue with the zip of the lambda source, but I run - **sometimes** into the same assume error you've got.  
 `The role defined for the function cannot be assumed by Lambda`. Run it a few times in a row without deleting the role (and ansible doesn't detect a change), it is accepted. Dunno what here the problem is?




I'm now able to run the integration test: https://gist.github.com/markuman/6a4e7bf57ce78900a4e025c73f8e556c  
But the `elb_target` test fails for me due some region missconfiguration.  
the `elb_lambda_target` succeeded too, but might fail with the iam_role thing described above.

jillr reviewed (changes requested):
@markuman Sorry for the confusion when I mentioned on irc I could run the tests from your branch - I didn't notice they weren't being included in full-test.yml, that was my fault.  Once you've got aws_connection_info added please make sure to include the new test role in test/integration/targets/elb_target/playbooks/full_test.yml.

jillr commented:
@markuman Got your irc message, your cloud-config-aws.ini should look like:
```
[default]
aws_access_key: AKIAAAAAAAAAAAAAAAA
aws_secret_key: XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
security_token: 
aws_region: us-east-1  # or whatever you're using as your default
```
Then assuming you have added elb_lambda_target to test/integration/targets/elb_target/playbooks/full_test.yml and that your test/integration/targets/elb_target/playbooks/roles/elb_lambda_target/tasks/main.yml looks like:
```
---
- name: set up aws connection info
  set_fact:
    aws_connection_info: &aws_connection_info
      aws_access_key: "{{ aws_access_key }}"
      aws_secret_key: "{{ aws_secret_key }}"
      security_token: "{{ security_token }}"
      region: "{{ aws_region }}"
  no_log: yes

- name: set up lambda as elb_target

  block:
    - name: Get the current caller identity facts
      aws_caller_facts:
        <<: *aws_connection_info
      register: caller_facts
## etc, etc for the rest of the aws tasks
```
You should then be able to execute the tests with `ansible-test integration --docker elb_target --allow-unsupported`.  I'm getting failures with the archive task though and didn't get that straightened out quickly.  For the sake of testing I pre-zipped ansible_lambda_target.py.zip and tried uploading that, but I'm getting this error:
```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: botocore.errorfactory.InvalidParameterValueException: An error occurred (InvalidParameterValueException) when calling the CreateFunction operation: The role defined for the function cannot be assumed by Lambda.
fatal: [localhost]: FAILED! => {"boto3_version": "1.9.130", "botocore_version": "1.12.188", "changed": false, "error": {"code": "InvalidParameterValueException", "message": "The role defined for the function cannot be assumed by Lambda."}, "msg": "Trying to create function: An error occurred (InvalidParameterValueException) when calling the CreateFunction operation: The role defined for the function cannot be assumed by Lambda.", "resource_actions": ["lambda:CreateFunction", "lambda:GetFunction"], "response_metadata": {"http_headers": {"connection": "keep-alive", "content-length": "90", "content-type": "application/json", "date": "Tue, 16 Jul 2019 20:25:31 GMT", "x-amzn-errortype": "InvalidParameterValueException", "x-amzn-requestid": "71eb4b48-3636-4210-9bc6-11b938d17409"}, "http_status_code": 400, "request_id": "71eb4b48-3636-4210-9bc6-11b938d17409", "retry_attempts": 0}}
```

Commit:
fix integration test

Commit:
use pause module to wait until the role is ready to use, remove hardcoded region

Commit:
enable elb_lambda_target test

markuman commented:
@jillr I fixed the integration test now. I need to wait 1 minute so that the role can be used.  
I also removed the hardcode region "eu-central-1", because the `elb_target` test must be run in region `us-east-1` or `us-east-2` due the defaults of the ami id.  
Both integration test succeeded now.

jillr reviewed (approved):
Tests are passing locally for me now, thanks for your work on this @markuman
